### ISSUE TITLE
out parameter in expression bodied method don't work as expected

### ISSUE THREAD

**Issue by FaustVX** [dotnet/roslyn#38243]
**Version Used**: C#8 .Net Core 3.0 preview 7

**Steps to Reproduce**:

``` csharp
public static string CreateShortSizedString(this ReadOnlyMemory<byte> span, out ReadOnlyMemory<byte> tail)
    => CreateSizedString(span, GetShort(span, out span), out tail);

public static short GetShort(this ReadOnlySpan<byte> span, out ReadOnlySpan<byte> tail)
    => new System.Buffers.Binary.BinaryPrimitives.ReadInt16BigEndian(span.Splice(2, out tail)));

public static string CreateSizedString(ReadOnlyMemory<byte> span, int length, out ReadOnlyMemory<byte> tail)
    => span.Splice(length, out tail).CreateString();
```

`span.Splice(length, out tail)` return the first segment of the array, and `out tail` is the 2nd segment.

VS show this warning `IDE0059 : Unnecessary assignment of a value to 'span'` for the `out span` of `GetShort`.

**Expected Behavior**:
After calling `GetShort`, `span` should be modified.
But on `CreateSizedString`, the span is not.

**Actual Behavior**:
When calling `CreateSizedString`, the span is the original Span.

`CreatehortSizedString` works fine written like this, but it should works also in the other way.
```csharp
public static string CreateShortSizedString(this ReadOnlySpan<byte> span, out ReadOnlySpan<byte> tail)
{
    var length = GetShort(span, out span);
    return CreateSizedString(span, length, out tail);
}
```
And also with ref like this (but I need to have the out parameter)
```csharp
public static string CreateShortSizedString(ref ReadOnlyMemory<byte> span)
    => CreateSizedString(ref span, GetShort(span, out span));

public static string CreateSizedString(ref ReadOnlyMemory<byte> span, int length)
    => span.Splice(length, out span).CreateString();
```

**Comment by gafter** [comment]
Arguments are evaluated in order.  Your original code took a copy of span and passed it as the first argument before modifying it.  So the behavior is as required by the language.

**Comment by FaustVX** [comment]
I don't find this very logical, because `GetShort` is called before `CreateSizedString`.
So the span should be modified before passing it in `CreateSizedString`

**Comment by gafter** [comment]
Arguments are evaluated left to right.  What order would you prefer?

**Comment by FaustVX** [comment]
Yes, the order is left to right, but `CreateSizedString `needs the value of `GetShort`.
So firstly, `GetShort` is called, then `CreateSizedString `is called, with left to right order, but `span` is already modified at this moment.

Also, when I write the second method,
```csharp
public static string CreateShortSizedString(this ReadOnlySpan<byte> span, out ReadOnlySpan<byte> tail)
{
    var length = GetShort(span, out span);
    return CreateSizedString(span, length, out tail);
}
```
VS show me an option to inline `length`, like in the first example.
For me, it makes sense that these 'light bulb' options are only for refactoring the code, not modifying the logic of my code (for that, I don't say it **should** work like that, but this is what I think, I could totally be wrong on that)

**Comment by gafter** [comment]
The first argument of `CreateSizedString` is evaluated before the second.

Do you know the difference between value types and reference types?

**Comment by FaustVX** [comment]
Yes I know the difference.
But I don't thought that the first argument is not evaluated as it's just a value, not a method call.
Now I'll know that. it's ok.

Also, about the refactoring options of my last comment, I am right or not ?
Because I don't think it's very useful to provide options which modify the logic of methods.

**Comment by jmarolf** [comment]
> Also, about the refactoring options of my last comment, I am right or not ?
I believe @CyrusNajmabadi has a well-reasoned rationale here

**Comment by CyrusNajmabadi** [comment]
> Yes, the order is left to right, but CreateSizedString needs the value of GetShort.

Yes.  But it computes that after figuring out the value of the first arg (as required by the language), which is not something that could possibly change as it would have massive implications for back compat.

**Comment by CyrusNajmabadi** [comment]
> VS show me an option to inline length, like in the first example.
For me, it makes sense that these 'light bulb' options are only for refactoring the code, not modifying the logic of my code

That's not the philosophy we take.  These options are just there to help speed you up and not force you to have to manually do something if you want to do it.  At no point does VS go out of it's way to push you to making this change.  It simply allows you to do it if you *want* to make this change.

When 'inlining' *anything* almost 99% of the time the semantics will change**.  So it's a definite *non-goal to preserve semantics with that feature as it would basically prevent the feature from ever being applicable in the first place.

--

i.e. the only time it wouldn't change would literally be if there was absolutely *no* code that ran between the computation of the local, and the very first reference of the local.  The vast majority of times people inline, that isn't the case at all.

**Comment by FaustVX** [comment]
Ok, thanks for the explanation.

**Comment by FaustVX** [comment]
I just found that using named argument to reorder the calling order of arguments of `CreateSizedString` in `CreateshortSizedString`, without modifying any method signature, all works as I originally expected.

Thank you all for the explanation.
